### PR TITLE
Setters

### PR DESCRIPTION
**Pull request by noamr**
* Refactor setters to generate simple bind code
* Allow deep setters when internal object/array doesn't exist
* Add push setter

### CODE REVIEW

**Comment by avi** [comment]
Arg0 arg1 etc can appear in the middle of the path not just in the end.
Setter(‘namespace’,arg0,’someKey’) should generate $setter.bind(null, (arg0,value) => set([‘namespace’, arg0, ‘someKey’], value))

**Comment by noamr** [comment]
> Arg0 arg1 etc can appear in the middle of the path not just in the end.
> Setter(‘namespace’,arg0,’someKey’) should generate $setter.bind(null, (arg0,value) => set([‘namespace’, arg0, ‘someKey’], value))

Of course, I think that use case should work... I'll add a test and make sure

**Comment by noamr** [comment]
> Arg0 arg1 etc can appear in the middle of the path not just in the end.
> Setter(‘namespace’,arg0,’someKey’) should generate $setter.bind(null, (arg0,value) => set([‘namespace’, arg0, ‘someKey’], value))

Of course, that's covered. Added a test

**Comment by noamr** [comment]
> Arg0 arg1 etc can appear in the middle of the path not just in the end.
> Setter(‘namespace’,arg0,’someKey’) should generate $setter.bind(null, (arg0,value) => set([‘namespace’, arg0, ‘someKey’], value))

Here it will be generated to $setter.bind(null, (...args) => set(['namespace', args[0], 'someKey'], value)

File: src/templates/optimizing.js [hunk 1]
```diff
@@ -928,6 +928,32 @@ function library() {
...
+      ensurePath(pathWithKey)
+      const key = pathWithKey[pathWithKey.length - 1]
+      const path = pathWithKey.slice(0, pathWithKey.length - 1)
+      const arr = getAssignableObject(path, path.length)
+      const origLength = arr.length;
+      const end = len === newItems.length ? key + len : Math.max(origLength, origLength + newItems.length - len);
+      for (let i = key; i < end; i++ ) {
+        triggerInvalidations(arr, i, true);
+      }
+      invalidatePath(path)
```
Comment (avi) [inline src/templates/optimizing.js]: invalidatePath with the path without the key will cause the entire array to be invalidated

Comment (noamr) [reply src/templates/optimizing.js]: Fixed. also added push setter

Comment (noamr) [reply src/templates/optimizing.js]: Oh shit yea I forgot the invalidates bool flag.... sec

Comment (noamr) [reply src/templates/optimizing.js]: It's fixed now.

**Pull request merged**